src/main/java/com/r307/arbitrader/service/cache/OrderVolumeCache.java

```diff
@@ -0,0 +1,85 @@
+package com.r307.arbitrader.service.cache;
+
+import org.knowm.xchange.Exchange;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Cache order volumes to avoid rate limiting. Order volumes don't change
+ * once they're placed so they're safe to cache.
+ */
+public class OrderVolumeCache {
+    public static final int CACHE_SIZE = 4;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OrderVolumeCache.class);
+
+    private final Map<String, BigDecimal> cache = new HashMap<>();
+    private final List<String> keys = new ArrayList<>();
+
+    /**
+     * Return a cached volume by exchange and order ID.
+     *
+     * @param exchange The exchange the order is on.
+     * @param orderId The order ID of the order.
+     * @return The volume of the order, if it is in the cache.
+     */
+    public BigDecimal getCachedVolume(Exchange exchange, String orderId) {
+        BigDecimal value = cache.get(computeCacheKey(exchange, orderId));
+
+        if (value == null) {
+            LOGGER.debug("Cache returned null for order {}:{}",
+                exchange.getExchangeSpecification().getExchangeName(),
+                orderId);
+
+            return null;
```

Comment:
I suggest to return optional instead of null. To avoid possible nasty NPE in the future

Reply:
Good idea. I'm converting all three caches to use `Optional`.